### DISCUSSION TITLE
Downloading/locating export?

### DISCUSSION THREAD

**Discussion by Lab-Things** [outline/outline#5984] [Self Hosting] [answered]
Hey all,

I'm trying to create a sort of emergency backup bash script that does the following (the script is running on the same server btw);
1. Export all documents to JSON
2. Upload the generated .zip file to an external server

For part 1 I use the following code;
```
response=$(curl -s -X POST -H "authorization: $TOKEN" -H "content-type: application/json" -H "accept: application/json" -d '{"format": "json"}' https://outline.mydomain.com/api/collections.export_all)
export_id=$(echo "$response" | jq -r '.data.fileOperation.id')
```

I've found that this successfully generates an export zip file. I can download it using the Outline interface.
The next step for me would be to locate the generated zip file, which is giving me some troubles. My first thought was to use the `fileOperations.redirect` API call. 

When I use this curl command 

```
curl -s -X POST -H "authorization: $TOKEN" -H "content-type: application/json" -H "accept: application/json" -d '{"id": "'$export_id'"}' https://outline.mydomain.com/api/fileOperations.redirect
```
The response is `Redirecting to https://loooooooongurl`. 
I can't for the life of me get curl to actually download this file, adding -L or -O just downloads a 76 byte empty file. I have no idea how I would go about locating this file otherwise.

Any pointers would be highly appreciated!

**Comment by tommoor** [comment]
This looks correct overall, your intuition of using the redirect endpoint is right. You would need the `-L` flag to have curl follow the location header that's returned.

What's in the 76 byte file?

**Reply by Lab-Things** [reply]
Thanks for responding! The 76 byte file is empty, its like executing `touch xxxx`. Glad to know I seem to be on the right track though!

**Reply by tommoor** [reply, marked as answer]
Seems like there is a bug here when using `POST` to access `fileOperations.redirect` it uses a `POST` for the subsequent download request, which fails as it should be a GET. The following works:

```
curl -L -o export.zip -v -s -X GET -H "authorization: $TOKEN" -H "content-type: application/json" "https://outline.mydomain.com/api/fileOperations.redirect?id=$export_id"
```

**Reply by Lab-Things** [reply]
You are a wizard. Thanks!!

**Reply by tommoor** [reply]
Quick followup – this wasn't a bug in Outline which correctly returns a `302` response, but using `-X POST` in curl forces it to do both the initial and the redirect as a POST for some reason. You can just remove that entirely and it works as well.

**Comment by amoralesc** [comment]
This is a god-send!! I was going mad trying to understand why cURL wasn't downloading the file export.

I'm adding the logs I was getting so anyone in the future that googles them gets pointed here.

```
<Error><Code>BadRequest</Code><Message>An error occurred when parsing the HTTP request POST at &#39;/bucket/uploads/...&#39;</Message><Resource>/bucket/uploads/...</Resource><RequestId></RequestId><HostId></HostId></Error>
```

And some keywords: Outline API document export download parsing error cURL

```
curl --request POST --url 'https://outline.domain.com/api/fileOperations.redirect?id=id' --header 'Authorization: Bearer token' --location
```

Redacted some content for privacy
